content/blog/towards-continuous-performance-regression-testing.asciidoc

```diff
@@ -0,0 +1,594 @@
+Based on https://openjdk.java.net/jeps/328[JDK Flight Recorder] events, JfrUnit allows you define and execute assertions e.g. against expected memory allocation, database I/O, or number of executed SQL statements, for a given workload.
+Starting off from a defined base line, future failures of such assertions are indicators for potential performance regressions in an application, as a code change may have introduced higher GC pressure,
+the retrieval of unneccessary data from the database, or SQL problems commonly induced by ORM tools, like N+1 SELECT statements.
+
+<!--more-->
+
+JfrUnit provides means of identifying and analyzing such anomalies in a reliable, environment independent way in standard JUnit tests,
+before they manifest as actual performance regressions in production.
+Test results are independent from wall clock time and thus provide actionable information also when not testing with production-like hardware and data volumes.
+
+This post is the first one in a series of two:
+
+* Part one provides an overview over continuous performance unit testing with JfrUnit,
+based on the examples of asserting expected memory allocation and database I/O
+* The second part will discuss the assertion of the SQL statements executed by an application,
+based on the instrumentation of JDBC APIs using https://wiki.openjdk.java.net/display/jmc/The+JMC+Agent[JMC Agent]
+
+== Getting Started With JfrUnit
+
+JfrUnit is an extension for JUnit 5, which integrates Flight Recorder into unit tests;
+it makes it straight forward to initiate a JFR recording for a given set of event types,
+execute some test routine, and then assert the JFR events which should have been produced.
+
+Here is a basic example of a JfrUnit test:
+
+[source,java]
+----
+@JfrEventTest // <1>
+public class JfrUnitTest {
+
+  public JfrEvents jfrEvents = new JfrEvents();
+
+  @Test
+  @EnableEvent("jdk.GarbageCollection") // <2>
+  @EnableEvent("jdk.ThreadSleep")
+  public void shouldHaveGcAndSleepEvents() throws Exception {
+    System.gc();
+    Thread.sleep(1000);
+
+    jfrEvents.awaitEvents(); // <3>
+
+    ExpectedEvent = event("jdk.GarbageCollection"); // <4>
+    assertThat(jfrEvents).contains(event); 
+    
+    ExpectedEvent = event("jdk.GarbageCollection") // <4>
+        .with("cause", "System.gc()"));
+    assertThat(jfrEvents).contains(event); 
+
+    ExpectedEvent = event("jdk.ThreadSleep").
+        with("time", Duration.ofSeconds(1)));
+    assertThat(jfrEvents).contains(event); 
+
+    assertThat(jfrEvents.ofType("jdk.GarbageCollection")).hasSize(1); // <5>
+  }
+}
+----
+<1> `@JfrEventTest` marks this as a JfrUnit test, activating its extension
+<2> All JFR event types to be recorded must be enabled via `@EnableEvent`
+<3> After running the test logic, `awaitEvents()` must be invoked as a synchronization barrier,
```

Comment:
so, design question.. why not expressing all expectations declaratively rather than have the user deal with the need for barriers?

Reply:
Yes, I was (am) considering that, as it makes the barrier issue go away, as you say. But declarative makes for much less flexibility in terms of expressing those expected values, you'd have to come up with a solution for summing up, aggregations, custom events etc. You'll land in "programming in annotations" land pretty quickly. So I'm leaning towards the imperative approach.

Reply:
ah yes I understand it better after having seen the second example.